katsdpimageutils/primarybeamcorrection/primary_beam_correction.py

```diff
@@ -0,0 +1,250 @@
+        ratio = rho/theta_b
+        num = np.cos(1.189*np.pi*ratio)
+        dem = 1-4*(1.189*ratio)**2
+        a_b = (num/dem)**2
+        flux_density.append(a_b)
+    return flux_density
+
+
+def beam_pattern(path):
+    """
+    Making beam pattern using the Cosine-squared power approximation from Mauch et al. (2020)
+    """
+    # Reading the fits file
+    data = read_fits(path)
+    # Getting pixel indices
+    row, col = np.indices((data.shape[2], data.shape[3]))
+    row = np.ravel(row)
+    col = np.ravel(col)
+    # Getting the central frequency of the image header given.
+    c_freq = central_freq(path)
+    # Getting radial separation between sources and the phase centre
+    phase_center, image_wcs = get_positions(path)
+    # Calling the radial_offset function
+    separation_rad = radial_offset(phase_center, image_wcs, row, col)
+    # Calling the cosine_power_pattern function
+    beam_list = cosine_power_pattern(separation_rad, c_freq)
+    return beam_list
+
+
+def mad(data):
+    """
+    Calculating median absolute deviation (MAD).
+    """
+    MAD_TO_SD = 1.4826
+    med = np.nanmedian(data)
+    dev = np.abs(data - med)
+    return med, MAD_TO_SD * np.nanmedian(dev)
+
+
+def standard_deviation(data):
+    """
+    Calculating the standard deviation of each frequency plane using MAD. Rejecting pixels more
+    than 5 sigma from the mean until either no more pixels are rejected or a maximum  of 50
+    iterations is reached.
+    Returns: Weights per each frequency plane
+    """
+    diff = 1
+    i = 0
+    data = data[data != 0.0]
+    if len(data) == 0:
+        return float(0.0), float(0.0)
+    med, sd = mad(data)
+    while diff > 0:
+        i += 1
+        if i > 50:
+            return 1/(sd)**2
+        old_sd = sd
+        old_len = len(data)
+        cut = np.abs(data - med) < 5.0 * sd
+        if np.all(~cut):
+            return 1/(sd)**2
+        data = data[cut]
+        med, sd = mad(data)
+        diff = old_len - len(data)
+        if sd == 0.0:
+            return 1/(old_sd)**2
+    return 1/(sd)**2
+
+
+def weighted_average(arr, weights, mask_image):
+    """
+    Computing weighted average of all the frequency planes.
+    """
+    masked_arr = np.ma.MaskedArray(arr, mask=mask_image)
+    wt_average = np.average(masked_arr, weights=weights, axis=0)
+    return wt_average.data
+
+
+def primary_beam_correction(beam_pattern, path):
+    """
+    Correcting for the primary beam effects.
+    """
+    raw_image = read_fits(path)
+    nterm = raw_image.header['NTERM']
+    snr = []
+    pbc_image = []
+    mask_image = []
+    # Getting all the values with attenuated flux of less than 10% of the peak
+    beam_mask = beam_pattern[-1] <= 0.1
+    for i in range(len((beam_pattern))):
+        beam_pattern[i][beam_mask] = np.nan
+        # Getting the rms noise in each image (before primary beam correction)
+        snr.append(standard_deviation(np.ravel(raw_image.data[0, i+nterm, :, :])))
```

Comment:
Is SNR the right term here? It seems like a measurement of noise but independent of the signal strength.

Reply:
It is the inverse variance, so I will rather just call it noise.